stroyprombeton/context.py

```diff
@@ -204,37 +192,90 @@ def context(self) -> typing.ContextDict:
         }
 
 
-# @todo #449:60m  Remove FetchPositions class.
-#  Create a new context class to search and filter positions.
-#  Move filter_options and slice_products methods to a created class.
-#  FilteredOptions created from `Catalog.filter_positions` is the example.
-class FetchPositions(Catalog):
+class SearchedOptions(context.Context):
 
     LOOKUPS = [
         'product__name__icontains',
         'code__icontains',
         'mark__icontains',
     ]
 
-    # redefined just to type hint
-    def __init__(self, request_data_: request_data.FetchProducts):
-        super().__init__(request_data_)
-
-    def filter_options(self, tags) -> stb_models.OptionQuerySet:
-        options = FilteredOptions(self.category, tags.qs(), self.request_data)
+    def __init__(
+        self,
+        options: stb_models.OptionQuerySet,
+        request_data_: request_data.FetchProducts
+    ):
+        self.options = options
+        self.request_data = request_data_
 
+    def qs(self) -> stb_models.OptionQuerySet:
         if self.request_data.filtered and self.request_data.term:
             return search(
                 self.request_data.term,
-                options.qs(),
+                self.options,
                 self.LOOKUPS,
                 ordering=('product__name', )
             )
         else:
-            return options.qs()
+            return self.options
 
-    def slice_products(self, options: stb_models.OptionQuerySet):
+    def context(self) -> typing.ContextDict:
+        return {
+            'positions': self.qs()
+        }
+
+
+class SlicedOptions(context.Context):
+    def __init__(
+        self,
+        options: stb_models.OptionQuerySet,
+        request_data_: request_data.FetchProducts
+    ):
+        self.options = options
+        self.request_data = request_data_
+
+    def qs(self) -> stb_models.OptionQuerySet:
         offset, limit = self.request_data.offset, self.request_data.length
-        return context.Products(
-            options[offset:offset + limit]
+        return self.options[offset:offset + limit]
+
+    def context(self) -> typing.ContextDict:
+        return {
+            'products': self.qs()
+        }
+
+
+class FetchPositions(context.Context):
+
+    # redefined just to type hint
+    def __init__(self, request_data_: request_data.FetchProducts):
+        self.request_data = request_data_
+
+    # TODO - this method to the new `context.Category` class
+    @property
+    def category(self):
+        return get_object_or_404(
+            stb_models.Category.objects.active().prefetch_related('page'),
+            id=self.request_data.id
         )
+
+    def context(self) -> typing.ContextDict:
+        tags = FilteredTags(stb_models.Tag.objects.all(), self.request_data)
+        # @todo #514:30m  Pass Context objects inside context.Options classes.
```

Comment:
It'd be nice to generalize this for every `Context` class but not only for `Options`